subprojects/ide/src/integTest/groovy/org/gradle/plugins/ide/AbstractSourcesAndJavadocJarsIntegrationTest.groovy

```diff
@@ -334,12 +339,191 @@ dependencies {
+            apply plugin: "idea"
+            apply plugin: "eclipse"
+
+            dependencies {
+                implementation gradleApi()
+            }
+            """
+
+        when:
+        succeeds ideTask
+
+        then:
+        TestFile sourcesDir = gradleDistributionSrcDir()
+        assertContainsGradleSources(sourcesDir)
+        ideFileContainsGradleApiWithSources("gradle-api", sourcesDir)
+    }
+
+    @ToBeFixedForInstantExecution
+    def "skips gradleApi() sources when not present and not available on remote server"() {
+        given:
+        requireGradleDistribution()
+
+        executer.withEnvironmentVars('GRADLE_REPO_OVERRIDE': "$server.uri/")
+        server.expectHeadMissing("/distributions-snapshots/gradle-${distribution.version.version}-src.zip")
+
+        buildScript """
+            apply plugin: "java"
+            apply plugin: "idea"
+            apply plugin: "eclipse"
+
+            dependencies {
+                implementation gradleApi()
+            }
+            """
+
         when:
         succeeds ideTask
 
         then:
-        ideFileContainsGradleApiWithSources("gradle-test-kit", sourcesDir.getPath())
-        ideFileContainsGradleApiWithSources("gradle-api", sourcesDir.getPath())
+        assertSourcesDirectoryDoesNotExistInDistribution()
+        ideFileContainsGradleApi("gradle-api")
+    }
+
+    @ToBeFixedForInstantExecution
+    def "skips gradleApi() sources when not present and remote distribution has unexpected directory structure"() {
+        given:
+        requireGradleDistribution()
+        executer.withEnvironmentVars('GRADLE_REPO_OVERRIDE': "$server.uri/")
+
+        givenSourceDistributionExistsOnRemoteServer("distributions-snapshots", distribution.version.version, createZip("gradle-src.zip") {
+            root {
+                foo {
+                    file("/src/main/java/org/gradle/Test.java")
+                    file("/src/main/java/org/gradle/Test2.java")
+                }
+            }
+        })
+
+        buildScript """
+            apply plugin: "java"
+            apply plugin: "idea"
+            apply plugin: "eclipse"
+
+            dependencies {
+                implementation gradleApi()
+            }
+            """
+
+        when:
+        succeeds ideTask
+
+        then:
+        assertSourcesDirectoryDoesNotExistInDistribution()
+        ideFileContainsGradleApi("gradle-api")
+    }
+
+    @ToBeFixedForInstantExecution
+    def "does not download gradleApi() sources when offline"() {
+        given:
+        requireGradleDistribution()
+        executer.withEnvironmentVars('GRADLE_REPO_OVERRIDE': "$server.uri/")
+
+        buildScript """
+            apply plugin: "java"
+            apply plugin: "idea"
+            apply plugin: "eclipse"
+
+            dependencies {
+                implementation gradleApi()
+            }
+
+            idea.module.downloadSources = false
+            eclipse.classpath.downloadSources = false
```

Comment:
❌ We should also test with `--offline` flag. This case is more about explicitly disallowing downloading sources.